File: backend/app/models/conversation.py
import uuid
from datetime import datetime
from app.utils.timezone import get_now_naive

from sqlalchemy import Column, DateTime, Enum, ForeignKey, String, Text, UniqueConstraint
from sqlalchemy.orm import relationship

from app.database import Base


class Conversation(Base):
    __tablename__ = "conversations"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    type = Column(Enum("direct", "group"), nullable=False)
    name = Column(String(255), nullable=True)
    avatar_url = Column(Text, nullable=True)
    created_by = Column(String(36), ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime, nullable=False, default=get_now_naive)

    creator = relationship("User", foreign_keys=[created_by])
    members = relationship(
        "ConversationMember",
        back_populates="conversation",
        cascade="all, delete-orphan",
    )
    messages = relationship(
        "Message",
        back_populates="conversation",
        order_by="Message.created_at",
    )
    calls = relationship("Call", back_populates="conversation")


class ConversationMember(Base):
    __tablename__ = "conversation_members"
    __table_args__ = (
        UniqueConstraint("conversation_id", "user_id", name="uq_conversation_user"),
    )

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    conversation_id = Column(
        String(36),
        ForeignKey("conversations.id", ondelete="CASCADE"),
        nullable=False,
    )
    user_id = Column(
        String(36),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    role = Column(Enum("admin", "member"), nullable=False, default="member")
    joined_at = Column(DateTime, nullable=False, default=get_now_naive)

    conversation = relationship("Conversation", back_populates="members")
    user = relationship("User", back_populates="conversation_memberships")

File: backend-go/models/conversation.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Conversation struct {
	ID          string    `gorm:"type:uuid;primaryKey" json:"id"`
	Type        string    `gorm:"size:50;not null" json:"type"`
	Name        *string   `gorm:"size:255" json:"name"`
	AvatarURL   *string   `gorm:"size:500" json:"avatar_url"`
	CreatedByID string    `gorm:"type:uuid;not null" json:"created_by"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`

	Members []ConversationMember `gorm:"foreignKey:ConversationID" json:"members,omitempty"`
}

func (c *Conversation) BeforeCreate(tx *gorm.DB) error {
	if c.ID == "" {
		c.ID = uuid.New().String()
	}
	return nil
}

type ConversationMember struct {
	ID             string    `gorm:"type:uuid;primaryKey" json:"id"`
	ConversationID string    `gorm:"type:uuid;not null;uniqueIndex:idx_conv_user" json:"conversation_id"`
	UserID         string    `gorm:"type:uuid;not null;uniqueIndex:idx_conv_user;index" json:"user_id"`
	Role           string    `gorm:"size:50;default:member" json:"role"`
	JoinedAt       time.Time `json:"joined_at"`

	User User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

func (cm *ConversationMember) BeforeCreate(tx *gorm.DB) error {
	if cm.ID == "" {
		cm.ID = uuid.New().String()
	}
	if cm.JoinedAt.IsZero() {
		cm.JoinedAt = time.Now()
	}
	return nil
}
